src/backend/distributed/planner/multi_logical_planner.c

```diff
@@ -158,6 +181,795 @@ MultiLogicalPlanCreate(Query *originalQuery, Query *queryTree)
+	 */
+	ErrorIfQueryNotSupported(originalQuery);
+
+	/*
+	 * In principle, we're first trying subquery pushdown planner. If it fails
+	 * to create a logical plan, continue with trying the single table
+	 * repartition subquery planning.
+	 */
+	subqueryPushdownError = ContainsUnsupportedSubquery(originalQuery,
+														plannerRestrictionContext);
+	if (!subqueryPushdownError)
+	{
+		multiQueryNode = SubqueryPushdownMultiPlanTree(originalQuery);
+	}
+	else if (subqueryPushdownError)
+	{
+		int singleRelationRepartitionSubquery =
+			SingleRelationRepartitionSubquery(originalQuery);
+
+		RangeTblEntry *subqueryRangeTableEntry = NULL;
+		Query *subqueryTree = NULL;
+		DeferredErrorMessage *repartitionQueryError = NULL;
+		List *subqueryEntryList = NULL;
+
+		/*
+		 * If not eligible for single relation repartition query, we should raise
+		 * subquery pushdown error.
+		 */
+		if (!singleRelationRepartitionSubquery)
+		{
+			RaiseDeferredErrorInternal(subqueryPushdownError, ERROR);
+		}
+
+		subqueryEntryList = SubqueryEntryList(queryTree);
+		subqueryRangeTableEntry = (RangeTblEntry *) linitial(subqueryEntryList);
+		Assert(subqueryRangeTableEntry->rtekind == RTE_SUBQUERY);
+
+		subqueryTree = subqueryRangeTableEntry->subquery;
+
+		repartitionQueryError = DeferErrorIfSubqueryNotSupported(subqueryTree);
+		if (repartitionQueryError)
+		{
+			RaiseDeferredErrorInternal(repartitionQueryError, ERROR);
+		}
+
+		/* all checks has passed, safe to create the multi plan */
+		multiQueryNode = MultiPlanTree(queryTree);
+	}
+
+	Assert(multiQueryNode != NULL);
+
+	return multiQueryNode;
+}
+
+
+/*
+ * SingleRelationRepartitionSubquery returns true if it is eligible single
+ * repartition query planning in the sense that:
+ *   - Each level of the subquery contains a single RTE
+ *   - Only a single RTE_RELATION exists
+ *
+ * Note that the caller should still call DeferErrorIfSubqueryNotSupported() to ensure
+ * that Citus supports the subquery. Also, this function is designed to run on the
+ * original query.
+ */
+static bool
+SingleRelationRepartitionSubquery(Query *queryTree)
+{
+	List *rangeTableList = queryTree->rtable;
+	RangeTblEntry *rangeTableEntry = NULL;
+	Query *subqueryTree = NULL;
+
+	/* don't allow joines on top of unions */
+	if (list_length(rangeTableList) > 1)
+	{
+		return false;
+	}
+
+	rangeTableEntry = linitial(rangeTableList);
+	if (rangeTableEntry->rtekind == RTE_RELATION)
+	{
+		return true;
+	}
+
+	if (rangeTableEntry->rtekind != RTE_SUBQUERY)
+	{
+		return false;
+	}
+
+	subqueryTree = rangeTableEntry->subquery;
+
+	return SingleRelationRepartitionSubquery(subqueryTree);
+}
+
+
+/*
+ * ContainsUnsupportedSubquery iterates on the query's subquery entry list
+ * and uses helper functions to check if we can push down subquery
+ * to worker nodes. These helper functions returns a deferred error out
+ * if we cannot push down the subquery.
+ */
+static DeferredErrorMessage *
+ContainsUnsupportedSubquery(Query *originalQuery,
```

Comment:
We don't have `SafeToPushDownSubquery ()` anymore, we've renamed it to `RestrictionEquivalenceForPartitionKeys()` in the previous PRs.

`ContainsUnsupportedSubquery()` already existed in the main branch, I wanted to be consistent with it. I'll now update it to `DeferErrorIfContainsUnsupportedSubquery()` to make it consistent with the other error functions. Later, we can discuss to rename all together to prefix with `Check..` or so.